### PR TITLE
revised - update createFinalHaplotypTable, add argument checks, allow snps to be detected in haplotypes with indels, fix removePrimer leaves 1 primer base behind

### PR DESCRIPTION
**Pull request by jemunro**

### CODE REVIEW

File: R/processReads.R [hunk 1]
```diff
@@ -32,22 +32,22 @@ demultiplexReads <- function(fastqFileFwd, fastqFileRev, barcodeFileFwd, barcode
   progressReport(detail=msg)
   while(length(sr1 <- yield(f1)) > 0){
```
Comment (lerch-a) [inline R/processReads.R:33]: I added yield(f2) here as well, so loop stop if either of it is 0

File: R/processReads.R [hunk 2]
```diff
@@ -32,22 +32,22 @@ demultiplexReads <- function(fastqFileFwd, fastqFileRev, barcodeFileFwd, barcode
...
+
+    if (length(sr1) != length(sr2)) {
+      warning('Fastq files have unequal lengths, it is likely one is incomplete')
+      if (length(sr1) < length(sr2))
+        sr2 <- sr2[seq_along(sr1)]
+      else 
+        sr1 <- sr1[seq_along(sr2)]
+    }
+    
+    if (length(sr1) == 0)
```
Comment (lerch-a) [inline R/processReads.R:44]: With addtitional change in the while loop not needed anymore.

File: R/processReads.R [hunk 3]
```diff
@@ -32,22 +32,22 @@ demultiplexReads <- function(fastqFileFwd, fastqFileRev, barcodeFileFwd, barcode
...
+      break
     countReads <- countReads + length(sr1)
     
     if(is.null(adapterFwd)){
```
Comment (lerch-a) [inline R/processReads.R:48]: if statement not needed, as adapter Fwd handling not yet implemented

File: R/processReads.R [hunk 4]
```diff
@@ -105,40 +105,46 @@ removePrimer <- function(fastqFileR1, fastqFileR2, outputFile, primerFwd, primer
   filteredReads <- 0
   while(length(sr1 <- yield(f1)) > 0){
     sr2 <- yield(f2)
+    stopifnot(length(sr1) == length(sr2))
```
Comment (lerch-a) [inline R/processReads.R:108]: Stop not required as check already before with a stop statemnt.

File: R/processReads.R [hunk 5]
```diff
@@ -105,40 +105,46 @@ removePrimer <- function(fastqFileR1, fastqFileR2, outputFile, primerFwd, primer
...
     sr1 <- sr1[index]
     sr2 <- sr2[index]
     
-    index <- vmatchPattern(primerRev, sread(sr2), max.mismatch=max.mismatch, with.indels=with.indels)
-    index <- unlist(lapply(index, length))==1
+    index <- which(vcountPattern(primerRev, sread(sr2), max.mismatch=max.mismatch, with.indels=with.indels) == 1)
     sr2 <- sr2[index]
     sr1 <- sr1[index]
     
     if(length(sr1)>1){
```
Comment (lerch-a) [inline R/processReads.R:119]: Should this not be >0? I changed that now.

File: R/processReads.R [hunk 6]
```diff
@@ -200,7 +217,8 @@ mergeAmpliconReads <- function(fastqFileR1, fastqFileR2, outputDir, mergePrefix=
     args <- paste("--fastq_mergepairs", fastqFileR1[i], "--reverse", fastqFileR2[i],
                   "--fastqout", outputFile, "--fastq_truncqual", 1, "--fastq_maxns", 0)
     Rvsearch:::.vsearchBin(args=args)
-    return(data.frame(numRead=NA_integer_, ReadFile=outputFile, stringsAsFactors=F))
+    numRead = length(Biostrings::readDNAStringSet(outputFile, format = 'FASTQ'))
```
Comment (lerch-a) [inline R/processReads.R:220]: I usually use ShortRead::readFastq or streamFastq to use a low memory option.

File: R/callHaplotype.R [hunk 7]
```diff
@@ -120,48 +121,70 @@ callHaplotype <- function(x, detectability=1/100, minHaplotypCoverage=3, minRepl
...
-
+  
   return(x)
 }
 
 createFinalHaplotypTable <- function(outputDir, sampleTable, markerTable, refSeq, snpList, postfix, 
                                      minHaplotypCoverage=3, minReplicate=2, 
-                                     detectability=1/100, minSampleCoverage=300){
+                                     detectability=1/100, minSampleCoverage=300,
+                                     maxDel = 0L, maxIns = 0L){
```
Comment (lerch-a) [inline R/callHaplotype.R:150]: Can wie add the maxDel and maxIns in a later commit. I want at least a different release for that.

File: R/processReads.R [hunk 8]
```diff
@@ -151,12 +157,22 @@ removePrimer <- function(fastqFileR1, fastqFileR2, outputFile, primerFwd, primer
...
   else
-    return(data.frame(numReadIn=totalReads, numReadOut=filteredReads, 
-                      FileR1=paste(outputFile, "_F.fastq.gz", sep=""), FileR2=paste(outputFile, "_R.fastq.gz", sep=""), 
+    return(data.frame(numReadIn=totalReads, numReadOut=filteredReads,
+                      FileR1=paste(outputFile, "_F.fastq.gz", sep=""), FileR2=paste(outputFile, "_R.fastq.gz", sep=""),
                       stringsAsFactors=F))
 }
 
-demultiplexByMarker <- function(sampleTable, markerTable, trimFilenameExt="R1\\.fastq.gz", progressReport=message){
+demultiplexByMarker <- function(sampleTable, markerTable, outDir, trimFilenameExt="R1\\.fastq.gz", progressReport=message){
```
Comment (lerch-a) [inline R/processReads.R:165]: I'll change outDir to outputDir to be consistent with other functions.

File: R/callHaplotype.R [hunk 9]
```diff
@@ -120,48 +121,70 @@ callHaplotype <- function(x, detectability=1/100, minHaplotypCoverage=3, minRepl
...
+  outFreqDir <- file.path(outputDir, "frequencyFiles")
+  dir.create(outFreqDir, recursive = T)
+  res <- lapply(markerTable$MarkerID, function(marker){
+    
+    outFreqFiles <- file.path(outFreqDir, marker)
+    dir.create(outFreqFiles)
+    
     samTab <- sampleTable[sampleTable$MarkerID == marker,]
-    potSNPLst <- snpLst[[marker]]
+    snpSet <- as.data.frame(snpList[[marker]], stringsAsFactors = FALSE)
```
Comment (lerch-a) [inline R/callHaplotype.R:179]: the snpList should be allready a data.frame. We need to fix this before.

File: R/callHaplotype.R [hunk 10]
```diff
@@ -177,12 +200,12 @@ createFinalHaplotypTable <- function(outputDir, sampleTable, markerTable, refSeq
     chimeraFilenames <- checkChimeras(clusterFilenames[,"RepresentativeFile"], method="vsearch")
     
     # create an overview table
-    overviewHap <- createHaplotypOverviewTable(allHaplotypesFilenames=fnAllSeq, 
-                                               clusterFilenames=clusterFilenames, chimeraFilenames=chimeraFilenames,
-                                               referenceSequence=refSeq[marker], snpList=potSNPLst)
+    overviewHap <- createHaplotypOverviewTable(allHaplotypesFilenames=fnAllSeq, clusterFilenames=clusterFilenames, 
+                                               chimeraFilenames=chimeraFilenames, referenceSequence=refSeq[marker],
+                                               snpSet=snpSet, maxDel = maxDel, maxIns = maxIns)
```
Comment (lerch-a) [inline R/callHaplotype.R:205]: Removed maxIns and maxDel. We add that at a later commit

File: R/callHaplotype.R [hunk 11]
```diff
@@ -192,23 +215,20 @@ createFinalHaplotypTable <- function(outputDir, sampleTable, markerTable, refSeq
...
+    snps <- unique(na.omit(overviewHap$snps[idx]))
+    if (length(snps) > 0)
+      names(snps) <- paste(marker, seq_along(snps), sep="-")
     overviewHap$FinalHaplotype[idx] <- names(snps)[match(overviewHap$snps[idx], snps)]
     overviewHap <- as.data.frame(overviewHap)
     if(devMode)
       write.table(overviewHap, file=file.path(outputDir, sprintf("HaplotypeOverviewTable_%s%s.txt", marker, postfix)), sep="\t")
     
     # getHaplotype sequence
-    if(!is.null(refSeq)){
```
Comment (lerch-a) [inline R/callHaplotype.R:204]: You can not delete this if statement it is important. I do not require that the refSeq is provided to the function.
This works only for non length polymorphic marker!!!

File: R/callHaplotype.R [hunk 12]
```diff
@@ -247,41 +267,39 @@ createFinalHaplotypTable <- function(outputDir, sampleTable, markerTable, refSeq
...
-    idx <- split(1:dim(haplotypesSample)[2], samTab$SampleName)
-    lst <- lapply(idx, function(i){
-      tab <- callHaplotype(haplotypesSample[,i, drop=F], minHaplotypCoverage=minHaplotypCoverage, 
-                           minReplicate=minReplicate, detectability=detectability, minSampleCoverage=minSampleCoverage, 
-                           reportBackground=T)
-      tab <- cbind(samTab[rep(i,each=dim(tab)[1]), c("SampleID","SampleName","MarkerID")], 
-                   Haplotype=rownames(tab), Reads=as.integer(tab))
-      colnames(tab) <- c("SampleID","SampleName","MarkerID","Haplotype","Reads")
-      rownames(tab) <- NULL
+    # idx <- split(1:dim(haplotypesSample)[2], samTab$SampleID)
```
Comment (lerch-a) [inline R/callHaplotype.R:276]: I think here you break the filtering based on the replicates. I double check that.

File: R/filterHaplotype.R [hunk 13]
```diff
@@ -36,7 +36,8 @@ checkChimeras <- function(representativesFile, method="vsearch", progressReport=
 }
 
 
-createHaplotypOverviewTable <- function(allHaplotypesFilenames, clusterFilenames, chimeraFilenames, referenceSequence=NULL, snpList){
+createHaplotypOverviewTable <- function(allHaplotypesFilenames, clusterFilenames, chimeraFilenames, referenceSequence,
+                                        snpSet, maxDel = 0L, maxIns = 0L){
```
Comment (lerch-a) [inline R/filterHaplotype.R:40]: Again remove maxDel and maxIns. Should be added in a different commit with a new release.

File: R/filterHaplotype.R [hunk 14]
```diff
@@ -117,56 +118,52 @@ createHaplotypOverviewTable <- function(allHaplotypesFilenames, clusterFilenames
...
-	    as.character(subseq(sread(sr1), start=n, width = 1))
-	  })
-	  # snps <- do.call(paste0, snps)
-	  snps <- apply(do.call(cbind, snps), 1, paste, collapse="")
-	  names(snps) <- as.character(id(sr1))
-	  
-	  overviewHap[names(snps),"snps"] <- snps
-  }else{
-  	overviewHap[,"snps"] <- ""
+  for (fn in repfile){
```
Comment (lerch-a) [inline R/filterHaplotype.R:125]: for loops are were slow in R and should not be used.

File: R/filterHaplotype.R [hunk 15]
```diff
@@ -117,56 +118,52 @@ createHaplotypOverviewTable <- function(allHaplotypesFilenames, clusterFilenames
...
+    sr1 <- readFasta(fn)
+    vec <- do.call(rbind, strsplit(as.character(id(sr1)), "_"))
+    at <- match(vec[,1], overviewHap$HaplotypesName)
+    clusterSize <- as.integer(vec[,2])
+    overviewHap$representatives[at] <- T
+    overviewHap$singelton[at[clusterSize>1]] <- F
   }
   
+  #  Indels
```
Comment (lerch-a) [inline R/filterHaplotype.R:134]: I can not accept the following changes in this commit. We discuss this for a later commit/release.

File: R/filterHaplotype.R [hunk 16]
```diff
@@ -117,56 +118,52 @@ createHaplotypOverviewTable <- function(allHaplotypesFilenames, clusterFilenames
...
+  overviewHap$indels <- F
+  hap_set <- as.character(with(overviewHap, HaplotypesName[representatives & ! singelton]))
+	haps <- readDNAStringSet(allHaplotypesFilenames)[hap_set]
```
Comment (lerch-a) [inline R/filterHaplotype.R:137]: Why do you read the allHaplotypesFilenames and not reuse the representative sequence reads? I also think that it disables the possibility of my version for length-polymorphic handling.
I'm happy to include your handling in future, but maybe as a different function. Let us discuss this after the workshop.

**Review by lerch-a — COMMENTED** [review]
Hi, I looked though the code and added some comments. I'll going to run the workshop data and check if the behaviour is still the same if I'm not sure. I hope to finish the merge process on Sunday.
Keep the changes (indel handling) which I do not accepted, for a later time-point. But also merge additional changes which I made during merging of the pull request back into you fork (are just small things). Be careful not to overwrite your indel handling modification.

File: R/callHaplotype.R (hunk 7 above)
Comment (jemunro) [reply R/callHaplotype.R:150]: Yes good idea

File: R/callHaplotype.R (hunk 12 above)
Comment (jemunro) [reply R/callHaplotype.R:276]: I may have, I don't quite understand how the filtering here works. Could you explain what you are filtering based on replicates?

File: R/filterHaplotype.R (hunk 14 above)
Comment (jemunro) [reply R/filterHaplotype.R:125]: Actually for loops are only slow when you are returning something from them, like a vector or a list.

If we compare:
```R
> system.time({
+     for (i in seq_len(1e6)) {
+         i + 1
+     }
+ })
   user  system elapsed 
  0.029   0.000   0.028 
```
and 
```R
> system.time({
+     lapply(seq_len(1e6), function(i){
+         i + 1
+     })
+ })
   user  system elapsed 
  1.028   0.013   1.040 
```
we can see the for loop is faster in this case.
However if we try to grow a vector in the for loop it is painfully slow:
```R
system.time({
+     v <- numeric()
+     for (i in seq_len(1e6)) {
+         v <- c(v, i + 1)
+     }
+ })
(takes over 1 minute)
```

File: R/processReads.R (hunk 5 above)
Comment (jemunro) [reply R/processReads.R:119]: yes, correct

File: R/processReads.R (hunk 6 above)
Comment (jemunro) [reply R/processReads.R:220]: I assumed that readDNAStringSet would be lower memory as it doesn't read the quality scores, but I haven't tested that.

Comment (lerch-a) [reply R/processReads.R:220]: Good point.

File: R/callHaplotype.R (hunk 12 above)
Comment (lerch-a) [reply R/callHaplotype.R:276]: Samples are split by haploytpename and haplotypes with less than minRelplicate occurense are filtered out as noise. eg if minReplicate=2 and you have duplicates then a haplotype has to be in both samples.
